modules/fx/arrow-fx/src/main/kotlin/arrow/fx/IO.kt

```diff
@@ -1024,22 +1013,22 @@ sealed class IO<out A> : IOOf<A> {
     override fun unsafeRunTimedTotal(limit: Duration): Option<A> = unsafeResync(this, limit)
   }
 
-  internal data class Bind<E, out A>(val cont: IO<E>, val g: (E) -> IO<A>) : IO<A>() {
-    override fun unsafeRunTimedTotal(limit: Duration): Option<A> = throw AssertionError("Unreachable")
+  internal data class Bind<A, E, out B, out E2 : E>(val cont: BIO<E, A>, val g: (A) -> BIO<E, B>) : BIO<E2, B>() {
```

Comment:
maybe dumb question, but wouldn't you need to provide a transformation to E2 somehow? 

```suggestion
  internal data class Bind<A, E, out B, out E2 : E>(val cont: BIO<E, A>, val g: (A) -> BIO<E2, B>) : BIO<E2, B>() {
```

Reply:
No, you are allowed to return an `BIO` that returns a *more specific* error.

Example where we go from `ErrorDomain` to `Error1`.

```
sealed class ErrorDomain {
  object Error1 : ErrorDomain()
  object Error2 : ErrorDomain()
}
val test: BIO<ErrorDomain, Int> = BIO.just(1)
val other: BIO<ErrorDomain.Error2, String> = BIO.just("other")
val result: BIO<ErrorDomain.Error2, String> = test.flatMap { other }
```